// apollo-link/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Observable = exports.fromError = exports.fromPromise = exports.toPromise = exports.makePromise = exports.createOperation = undefined;

var _link = require('./link');

Object.keys(_link).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _link[key];
    }
  });
});

var _linkUtils = require('./linkUtils');

Object.defineProperty(exports, 'createOperation', {
  enumerable: true,
  get: function get() {
    return _linkUtils.createOperation;
  }
});
Object.defineProperty(exports, 'makePromise', {
  enumerable: true,
  get: function get() {
    return _linkUtils.makePromise;
  }
});
Object.defineProperty(exports, 'toPromise', {
  enumerable: true,
  get: function get() {
    return _linkUtils.toPromise;
  }
});
Object.defineProperty(exports, 'fromPromise', {
  enumerable: true,
  get: function get() {
    return _linkUtils.fromPromise;
  }
});
Object.defineProperty(exports, 'fromError', {
  enumerable: true,
  get: function get() {
    return _linkUtils.fromError;
  }
});

var _zenObservableTs = require('zen-observable-ts');

var _zenObservableTs2 = _interopRequireDefault(_zenObservableTs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.Observable = _zenObservableTs2.default;
