// apollo-link/dist/src/bundle.umd.js
'use strict';

var _typeof8 = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _typeof7 = typeof Symbol === "function" && _typeof8(Symbol.iterator) === "symbol" ? function (obj) {
	return typeof obj === "undefined" ? "undefined" : _typeof8(obj);
} : function (obj) {
	return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof8(obj);
};

var _typeof6 = typeof Symbol === "function" && _typeof7(Symbol.iterator) === "symbol" ? function (obj) {
	return typeof obj === "undefined" ? "undefined" : _typeof7(obj);
} : function (obj) {
	return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof7(obj);
};

var _typeof5 = typeof Symbol === "function" && _typeof6(Symbol.iterator) === "symbol" ? function (obj) {
	return typeof obj === "undefined" ? "undefined" : _typeof6(obj);
} : function (obj) {
	return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof6(obj);
};

var _typeof4 = typeof Symbol === "function" && _typeof5(Symbol.iterator) === "symbol" ? function (obj) {
	return typeof obj === "undefined" ? "undefined" : _typeof5(obj);
} : function (obj) {
	return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof5(obj);
};

var _typeof3 = typeof Symbol === "function" && _typeof4(Symbol.iterator) === "symbol" ? function (obj) {
	return typeof obj === "undefined" ? "undefined" : _typeof4(obj);
} : function (obj) {
	return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof4(obj);
};

var _typeof2 = typeof Symbol === "function" && _typeof3(Symbol.iterator) === "symbol" ? function (obj) {
	return typeof obj === "undefined" ? "undefined" : _typeof3(obj);
} : function (obj) {
	return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof3(obj);
};

var _typeof = typeof Symbol === "function" && _typeof2(Symbol.iterator) === "symbol" ? function (obj) {
	return typeof obj === "undefined" ? "undefined" : _typeof2(obj);
} : function (obj) {
	return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj === "undefined" ? "undefined" : _typeof2(obj);
};

(function (global, factory) {
	(typeof exports === 'undefined' ? 'undefined' : _typeof(exports)) === 'object' && typeof module !== 'undefined' ? factory(exports) : typeof define === 'function' && define.amd ? define(['exports'], factory) : factory(global.apolloLink = {});
})(undefined, function (exports) {
	'use strict';

	function __export(m) {
		for (var p in m) {
			if (!exports.hasOwnProperty(p)) exports[p] = m[p];
		}
	}
	Object.defineProperty(exports, "__esModule", { value: true });
	var apollo_link_core_1 = require("apollo-link-core");
	var apollo_link_batch_1 = require("apollo-link-batch");
	exports.BatchLink = apollo_link_batch_1.default;
	var apollo_link_batch_http_1 = require("apollo-link-batch-http");
	exports.BatchHttpLink = apollo_link_batch_http_1.default;
	var apollo_link_dedup_1 = require("apollo-link-dedup");
	exports.DedupLink = apollo_link_dedup_1.default;
	var apollo_link_http_1 = require("apollo-link-http");
	exports.HttpLink = apollo_link_http_1.default;
	var apollo_link_retry_1 = require("apollo-link-retry");
	exports.RetryLink = apollo_link_retry_1.default;
	var apollo_link_set_context_1 = require("apollo-link-set-context");
	exports.SetContextLink = apollo_link_set_context_1.default;
	var apollo_link_polling_1 = require("apollo-link-polling");
	exports.PollingLink = apollo_link_polling_1.default;
	var apollo_link_ws_1 = require("apollo-link-ws");
	exports.WebSocketLink = apollo_link_ws_1.default;
	__export(require("apollo-link-core"));
	exports.default = apollo_link_core_1.default;

	Object.defineProperty(exports, '__esModule', { value: true });
});
